src/Components/Components/src/ComponentResolver.cs

```diff
@@ -13,61 +14,40 @@ namespace Microsoft.AspNetCore.Components
     /// </summary>
     internal static class ComponentResolver
     {
-        /// <summary>
-        /// Lists all the types
-        /// </summary>
-        /// <param name="appAssembly"></param>
-        /// <returns></returns>
+        private static readonly string ComponentAssemblyName = typeof(IComponent).Assembly.FullName;
+        private static readonly ConcurrentDictionary<Assembly, Type[]> Cache =
+            new ConcurrentDictionary<Assembly, Type[]>();
+
         public static IEnumerable<Type> ResolveComponents(Assembly appAssembly)
         {
-            var componentsAssembly = typeof(IComponent).Assembly;
+            if (Cache.TryGetValue(appAssembly, out var resolvedComponents))
+            {
+                return resolvedComponents;
+            }
 
-            return EnumerateAssemblies(appAssembly.GetName(), componentsAssembly, new HashSet<Assembly>(new AssemblyComparer()))
-                .SelectMany(a => a.ExportedTypes)
-                .Where(t => typeof(IComponent).IsAssignableFrom(t));
+            var components = DiscoverComponents(appAssembly);
+            Cache.TryAdd(appAssembly, components);
+            return components;
         }
 
-        private static IEnumerable<Assembly> EnumerateAssemblies(
-            AssemblyName assemblyName,
-            Assembly componentAssembly,
-            HashSet<Assembly> visited)
+        private static Type[] DiscoverComponents(Assembly assembly)
```

Comment:
Added caching + requested change to the discovery in ComponentResolver. Might be helpful to view this file independently.